File: dec8/two.mjs
import fs from "fs/promises";

(async () => {
    const file = (await fs.readFile("./input.txt", "utf-8")).trim();
    let lines = file.split("\n");
    const map = new Map();
    lines.forEach((line, i) => {
        line.split("").forEach((char, j) => {
            if(char !== "."){
                map.get(char) ? map.get(char).push([j, i]) : map.set(char, [[j, i]]);
            }
        });
    });
    const set = new Set();
    const max_x = lines[0].length;
    const max_y = lines.length;
    for(let char of map){
        const entry = char[1];
        for(let i = 0; i < entry.length; i++){
            for(let j = 0; j < entry.length; j++){
                if(i !== j){
                    let dx = entry[i][0] - entry[j][0];
                    let dy = entry[i][1] - entry[j][1];
                    let px = entry[i][0];
                    let py = entry[i][1];
                    while(px >= 0 && px < max_x && py >= 0 && py < max_y){
                        set.add(`${px},${py}`);
                        px += dx;
                        py += dy;
                    }
                    while(px >= 0 && px < max_x && py >= 0 && py < max_y){
                        set.add(`${px},${py}`);
                        px -= dx;
                        py -= dy;
                    }
                }
            }
        }
    }
    console.log(set.size);
})()

File: dec8/two.rs
use std::fs;
use std::collections::{HashMap, HashSet};

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let content = fs::read_to_string("./input.txt")?;
    let file = content.trim();
    let lines: Vec<&str> = file.split('\n').collect();
    let mut map: HashMap<char, Vec<(usize, usize)>> = HashMap::new();

    for (i, line) in lines.iter().enumerate() {
        for (j, ch) in line.chars().enumerate() {
            if ch != '.' {
                map.entry(ch).or_insert_with(Vec::new).push((j, i));
            }
        }
    }

    let mut set: HashSet<String> = HashSet::new();
    let max_x = lines[0].len();
    let max_y = lines.len();

    for (_ch, entries) in &map {
        for i in 0..entries.len() {
            for j in 0..entries.len() {
                if i != j {
                    let dx = entries[i].0 as isize - entries[j].0 as isize;
                    let dy = entries[i].1 as isize - entries[j].1 as isize;
                    let mut px = entries[i].0 as isize;
                    let mut py = entries[i].1 as isize;

                    while px >= 0 && px < max_x as isize && py >= 0 && py < max_y as isize {
                        set.insert(format!("{},{}", px, py));
                        px += dx;
                        py += dy;
                    }

                    px = entries[i].0 as isize;
                    py = entries[i].1 as isize;

                    while px >= 0 && px < max_x as isize && py >= 0 && py < max_y as isize {
                        set.insert(format!("{},{}", px, py));
                        px -= dx;
                        py -= dy;
                    }
                }
            }
        }
    }

    println!("{}", set.len());

    Ok(())
}